presto-main/src/main/java/io/prestosql/operator/PagesIndex.java

```diff
@@ -84,8 +85,10 @@
     private final List<Type> types;
     private final LongArrayList valueAddresses;
     private final ObjectArrayList<Block>[] channels;
+    private final IntArrayList positionCounts;
```

Comment:
Seems like a shame to add this memory overhead in all cases when in practice it's only necessary for empty `channels`. Maybe we only use this when `channels.length == 0`? It'll definitely make the code uglier but it seems maybe worthwhile to avoid the extra memory penalty when it can be avoided.

Reply:
I know, but it shouldn't be much overhead -- roughly an int per page.